src/ProducerEventProcessor/UploaderCaptureEventCollector.cpp

```diff
@@ -0,0 +1,116 @@
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+UploaderCaptureEventCollector::UploaderCaptureEventCollector() {
+  // Create an output stream which will be used to convert capture events into parts of a well
+  // formatted capture file. This should never fail.
+  output_stream_ =
+      orbit_capture_file::CaptureFileOutputStream::Create(&capture_data_buffer_stream_);
+}
+
+UploaderCaptureEventCollector::~UploaderCaptureEventCollector() {
+  LOG("Total number of events uploaded: %u", total_uploaded_event_count_);
+  LOG("Total number of bytes uploaded: %u", total_uploaded_data_bytes_);
+  LOG("Total number of write event errors: %u", total_write_error_count_);
+
+  if (total_uploaded_event_count_ > 0) {
+    float average_bytes = static_cast<float>(total_uploaded_data_bytes_) /
+                          static_cast<float>(total_uploaded_event_count_);
+    LOG("Average number of bytes per event: %.2f", average_bytes);
+  }
+}
+
+void UploaderCaptureEventCollector::Stop() {
+  absl::MutexLock lock{&mutex_};
+
+  CHECK(output_stream_ != nullptr);
+  auto close_result = output_stream_->Close();
+  if (close_result.has_error()) {
+    LOG("Error while closing output steam: %s", close_result.error().message());
+  }
+}
+
+void UploaderCaptureEventCollector::AddEvent(ClientCaptureEvent&& event) {
+  {
+    absl::MutexLock lock{&mutex_};
+
+    // The output stream will be closed while processing the capture finish event. Drop events
+    // received after closing the output stream.
+    CHECK(output_stream_ != nullptr);
+    if (!output_stream_->IsOpen()) return;
+
+    auto write_result = output_stream_->WriteCaptureEvent(event);
+    if (write_result.has_error()) {
+      total_write_error_count_++;
+      return;
+    }
+    buffered_event_count_++;
+    buffered_event_bytes_ += event.ByteSizeLong();
+  }
+
+  // Close output stream after processing the capture finish event.
+  if (event.event_case() == ClientCaptureEvent::kCaptureFinished) Stop();
+}
+
+DataReadiness UploaderCaptureEventCollector::GetDataReadiness() const {
+  absl::MutexLock lock(&mutex_);
+
+  if (!capture_data_to_upload_.empty()) {
+    return DataReadiness::kHasData;
+
+  } else if (output_stream_->IsOpen() || buffered_event_bytes_ > 0) {
+    return DataReadiness::kWaitingForData;
+
+  } else {
+    return DataReadiness::kEndOfData;
+  }
+}
+
+void UploaderCaptureEventCollector::RefreshUploadDataBuffer() {
+  // Clear capture_data_to_upload_ immediately.
+  capture_data_to_upload_.clear();
+
+  // Refill capture_data_to_upload_ when there is enough data to upload.
+  mutex_.LockWhen(absl::Condition(
+      +[](UploaderCaptureEventCollector* self) ABSL_EXCLUSIVE_LOCKS_REQUIRED(self->mutex_) {
+        constexpr int kUploadEventCountInterval = 5000;
+        return self->buffered_event_count_ >= kUploadEventCountInterval ||
+               !self->output_stream_->IsOpen();
+      },
+      this));
```

Comment:
Regarding my last comment: I might not have been super-clear. Sry. I didn't mean to just remove the timeout. I meant we probably shouldn't wait here at all.

So my suggestion:
1. We lock the mutex using lock_guard
2. We check whether enough events have been queued or if the stream was closed.
3. If both is false we just return, which means `capture_data_to_upload_` will be empty
4. If not we take the buffer from `capture_data_buffer_stream_`.

Would this work? I'm basically asking why you decided to put the blocking lock in here. Is there a reason that I don't see?

Reply:
We might still need to use `LockWhen`, as the uploader will only request for refresh for once. 

After uploading data in upload data buffer (i.e., `capture_data_to_upload_`), the uploader calls `RefreshUploadDataBuffer` and pause the curl session. Then the collector keeps tell uploader `DataReadiness::kWaitForData` unless `capture_data_to_upload_` contains new data and is not empty anymore.

If we return an empty `capture_data_to_upload_`, the curl session will keep being paused and there won't be a second request for `RefreshUploadDataBuffer`.

Reply:
Ah okay. That means the API is missing a signal. We need to tell curl when new data is available after it has been paused.

Waiting for the Lock certainly "fixes" that, but what happens when the LockWhen call takes multiple seconds to finish? That would mean we are stuck for multiple seconds in the read callback and the HTTP session might just time out aborting the whole upload. Sure that's not likely to happen but from a design perspective we could easily fix that.

I think we should not block here and pause the connection instead and then resume uploading in the CURLOPT_PROGRESSFUNCTION callback as described in https://curl.se/libcurl/c/curl_easy_pause.html. WDYT?

Reply:
We have the signal to tell curl when data is ready. And we are using `XferInfoCallback`, which is similar as the `PROGRESSFUNCTION` callback you mentioned, to resume the HTTP session.